cost-center refactor

tobi-or-not-tobi opened the pull request:
- Introduce new UX
- move to separate lib
  - move components
  - move css
  - move i18n
- use child routes
- leverage split-view component
- leverage cx-table component
- refactor components (keep it small and clean, no inheritance) 
- refactor services
- keep URL navigation simple, so you can use it as a directory structure
- simplify initial i18n
- use reusable testing modules

closes #7861

tobi-or-not-tobi committed:
move routing to feature lib, introduce child routes

tobi-or-not-tobi committed:
fix npe

tobi-or-not-tobi committed:
fix missing routermodule

tobi-or-not-tobi committed:
leverage table and split-view

tobi-or-not-tobi committed:
normalize boolean flags

tobi-or-not-tobi committed:
introduce feature lib styles

tobi-or-not-tobi committed:
add icons

tobi-or-not-tobi committed:
organize i18n

tobi-or-not-tobi committed:
finalize base list compoent

tobi-or-not-tobi committed:
organisation page template styles

tobi-or-not-tobi committed:
Move i18n labels to feature lib

tobi-or-not-tobi committed:
improve organization styles

tobi-or-not-tobi committed:
add budget

tobi-or-not-tobi committed:
intermediate state

tobi-or-not-tobi committed:
fix broken breadcrumbs (from #8105)

tobi-or-not-tobi committed:
Merge branch 'feature/my-company' into feature/GH-cost-center-ui

tobi-or-not-tobi committed:
css fixes

dunqan committed:
idea to fix infinite pagination

dunqan committed:
Merge remote-tracking branch 'origin/feature/GH-cost-center-ui' into feature/GH-cost-center-ui

dunqan committed:
Revert "idea to fix infinite pagination"

This reverts commit b856318a

tobi-or-not-tobi committed:
start testing

tobi-or-not-tobi committed:
Merge branch 'feature/GH-cost-center-ui' of https://github.com/SAP/spartacus into feature/GH-cost-center-ui

znikola committed:
added declarations

tobi-or-not-tobi committed:
refactored listing services

tobi-or-not-tobi committed:
fix responsive structure

tobi-or-not-tobi committed:
Introduce test modules and aggregate them in a single shared testing module for organization

tobi-or-not-tobi committed:
fix tests

tobi-or-not-tobi committed:
Use cxRoute everwhere

tobi-or-not-tobi committed:
fix split view

tobi-or-not-tobi committed:
fix hidden first view when 3rd level cx-view is opened

tobi-or-not-tobi committed:
fix template reference

Platonn commented on projects/storefrontlib/src/shared/components/table/table.model.ts:
```diff
@@ -18,27 +17,42 @@ export interface Table<T = any> {
+  pagination?: PaginationModel;
 }
 
+// export interface TablePagination {
```
leftover?

Platonn commented on projects/storefrontlib/src/shared/components/table/table.service.ts:
```diff
@@ -131,13 +131,24 @@ export class TableService {
+        ? tableConfig.find((config) => config.breakpoint === bestMatch)
+        : {}),
+    };
+    // return bestMatch
```
leftover?

Platonn commented on projects/storefrontapp/src/styles.scss:
```diff
@@ -4,3 +4,8 @@
+@import 'my-account';
+// @import '@spartacus/my-account';
+
+// @import 'my-account/organization/index';
```
leftovers?

Platonn commented on projects/assets/src/translations/translation-chunks-config.ts:
```diff
@@ -73,17 +73,20 @@ export const translationChunksConfig: TranslationChunksConfig = {
+    // 'costCenterBudgets',
+    // 'costCenterAssignBudgets',
+
     'breadcrumbs',
```
`breadcrumbs.XXX` key sounds quite general. Why do we place it in the `organization.json` chunk?

dunqan replied:
I guess we are moving away from it, as Tobias explained the breadcrumb mechanism was not well designed. 
So all of it will be ideally removed like costCenter stuff.

tobi-or-not-tobi replied:
Organisation in this chunk will varnish when we move all to feature lib. And breadcrumbs should not be driven by localisations as much as possible.

Platonn commented on feature-libs/my-account/organization/src/styles/_section.scss:
```diff
@@ -0,0 +1,67 @@
+      }
+    }
+    .property {
+      // display: flex;
```
there are many commented code leftovers across in css files

tobi-or-not-tobi replied:
we can clean them up

Platonn commented on feature-libs/my-account/organization/src/components/shared/organization-list.service.ts:
```diff
@@ -0,0 +1,102 @@
+@Injectable({
+  providedIn: 'root',
+})
+export abstract class BaseOrganizationListService<T, P = PaginationModel> {
```
Can we align the class name with the file name for easier search in IDE?

Platonn commented on feature-libs/my-account/organization/src/components/cost-center/list/cost-center-list.service.ts:
```diff
@@ -0,0 +1,70 @@
+   * Populates the cost center data to a convenient table data model, so that we
+   * can skip specific conversion in the view logic where possible.
+   */
+  protected populateCostCenters({
```
`populate` vs. `normalize` (or `convert`)?

tobi-or-not-tobi replied:
I'm open to any name here;)

Platonn commented on feature-libs/my-account/organization/src/components/cost-center/form/cost-center-form.module.ts:
```diff
@@ -9,7 +9,7 @@ import {
   UrlModule,
 } from '@spartacus/core';
-import { FormErrorsModule } from '@spartacus/storefront';
+import { FormErrorsModule } from 'projects/storefrontlib/src/shared/components/form/form-errors/form-errors.module';
```
Since FormErrorsModule lives in a different lib, it should be imported from `@spartacus/strorefrontlib`

Platonn commented on feature-libs/my-account/organization/src/components/cost-center/budgets/list/cost-center-budget-list.component.ts:
```diff
@@ -0,0 +1,26 @@
+  );
+
+  constructor(
+    protected router: ActivatedRoute,
```
`router` vs `route`?

Platonn commented on projects/storefrontlib/src/shared/components/table/table.model.ts:
```diff
@@ -18,27 +17,42 @@ export interface Table<T = any> {
    * created however.
    */
-  data$: Observable<T[]>;
+  data: T[];
```
The JSDoc above says `The table data is provided in an observable, so that we can observe changes and benefit from change detection`. We should change the doc.

Platonn committed:
align filename with class BaseOrganizationListService

Platonn committed:
rename router to route for ActvatedRoute dependency

Pucek9 committed:
fix unit test

Pucek9 committed:
fix component test

Pucek9 committed:
add types

Pucek9 committed:
Merge branch 'feature/my-company' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-cost-center-ui

dunqan committed:
make app buildable

dunqan committed:
rename methods in costSenter facade

dunqan committed:
method rename

dunqan committed:
cost cemter config refactor

dunqan committed:
fix prod build css generation

dunqan committed:
Rename buildRandomStructure to buildFallbackStructure in a table service

Platonn commented on projects/assets/src/translations/translation-chunks-config.ts:
```diff
@@ -1,6 +1,4 @@
-interface TranslationChunksConfig {
```
This interface was redundant in the assets lib on purpose - not to have a dependency between `@spartacus/assets` and `@spartacus/core`

dunqan committed:
scss fix

Platonn committed:
test: improve modules with testing utils (#8153)

1. Dropped the OrganizationTestingModule, because the internal modules were not always used by the consuming test suites:
   - cx-pagination was not used in most cases
   - cx-table or cx-split-view or cx-view was used in half of cases (in various combinations)
   - cx-icon, cxUrl pipe and angular [routerLink] were used almost always
   - cxTranslate pipe was used always

2. Made all new testing modules private API (can be imported in spec filess across libraries).

3. Summary of modules imported in tests:
   -  I18nTestingModule, // public MOCK - already existing
   - UrlTestingModule, // private MOCK - quite smart (complex?), now supports many (but not all) use cases of cxUrl pipe
   - SplitViewTestingModule, // private MOCK - only < ng-content > for cx-view and cx-split-view (no logic of hiding/showing `cx-views`)
   - TableModule, // ORIGINAL - to allow for using embedded outlets
   - PaginationTestingModule, // private MOCK = ORIGINAL component + provide ORIGINAL default `PaginationConfig`
   - IconTestingModule, // private MOCK - simply renders the icon type

Platonn commented on projects/storefrontstyles/scss/components/organization/_organization-form.scss:
```diff
@@ -1,11 +1,11 @@
-      width: var(--cx-max-width, 75%);
-    }
-  }
+  // text-align: center;
```
Should it be removed?

Platonn committed:
remove obsolete commented out code (#8162)

dunqan committed:
feat: Cost center form service usage refactor (#8154)

move form ownership to form service

Pucek9 committed:
Merge branch 'feature/my-company' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-cost-center-ui

# Conflicts:
#	feature-libs/my-account/organization/src/components/cost-center/budgets/cost-center-budgets.component.html
#	feature-libs/my-account/organization/src/components/cost-center/budgets/cost-center-budgets.component.ts
#	projects/storefrontlib/src/cms-components/misc/icon/fontawesome-icon.config.ts
#	projects/storefrontlib/src/cms-components/misc/icon/icon.model.ts
